x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/CopySign.java

```diff
@@ -0,0 +1,159 @@
+        source().writeTo(out);
+        out.writeNamedWriteable(children().get(0));
+        out.writeNamedWriteable(children().get(1));
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, CopySign::new, children().get(0), children().get(1));
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        if (newChildren.size() != 2) {
+            throw new EsqlIllegalArgumentException("Function [{}] expects exactly two arguments, got [{}]", NAME, newChildren.size());
+        }
+        return new CopySign(source(), newChildren.get(0), newChildren.get(1));
+    }
+
+    @Override
+    public DataType dataType() {
+
+        if (dataType == null) {
+            resolveType();
+        }
+        return dataType;
+    }
+
+    @Override
+    public TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+        var magnitude = children().get(0);
+        var sign = children().get(1);
+        if (magnitude.dataType().isNumeric() == false) {
+            return new TypeResolution("Magnitude must be a numeric type");
+        }
+        if (sign.dataType().isNumeric() == false) {
+            return new TypeResolution("Sign must be a numeric type");
+        }
+        // TODO(pabloem): Looks like we don't currently support generalizing types
+        // when two types are incompatible (e.g. copySign(int, long) -> long instead of
+        // copySign(int, long) => ERROR).
+        dataType = magnitude.dataType();
+        TypeResolution resolution = TypeResolutions.isType(
+            magnitude,
+            t -> t == dataType,
+            sourceText(),
+            TypeResolutions.ParamOrdinal.fromIndex(1),
+            magnitude.dataType().typeName()
+        );
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    @Override
+    public boolean foldable() {
+        return Expressions.foldable(children());
+    }
+
+    @Override
+    public EvalOperator.ExpressionEvaluator.Factory toEvaluator(ToEvaluator toEvaluator) {
+        var dataType = dataType();
+        EvalOperator.ExpressionEvaluator.Factory magnitudeEvaluatorFactory = toEvaluator.apply(children().get(0));
+        EvalOperator.ExpressionEvaluator.Factory signEvaluatorFactory = toEvaluator.apply(children().get(1));
+        if (dataType == DataType.FLOAT) {
+            return new CopySignFloatEvaluator.Factory(source(), magnitudeEvaluatorFactory, signEvaluatorFactory);
+        } else if (dataType == DataType.DOUBLE) {
+            return new CopySignDoubleEvaluator.Factory(source(), magnitudeEvaluatorFactory, signEvaluatorFactory);
+        } else {
```

Comment:
FWIW we've been making a `private static final Map<DataType, BiFunction<Factory, Factory Factory>` with these signatures and using that for this. We also use it for the "is this a valid type signature" for the method. And it feels very java to have a map of builders for factories for evaluators. Brings me back to my spring BeanFactoryFactoryStrategyFactory roots.